### ISSUE TITLE
Gradients with respect to network and component parameters

### ISSUE THREAD

**Issue by zhui99** [elpetros99/ElectroDynJAX#1]
## Description

  Creating a `Network` with recent versions of Equinox produces the following warning:

  ```text
  UserWarning: Using `field(init=False)` on `equinox.Module` can lead to surprising behaviour when used around `jax.grad`.

  In the following example, observe how JAX computes gradients with respect to the derived `.len` attribute, which is a PyTree leaf passed across the
  `jax.grad` boundary, and that there are no gradients with respect to the original `.a` or `.b` attributes.
```

The warning is emitted when Network(...) is constructed, for example at: time_domain_differentiable/cli/main_clean_rpnn_IEEE24.py:848

 ## Environment

  Python: 3.11
  JAX: 0.10.2
  jaxlib: 0.10.2
  Equinox: 0.13.8

I tested that the gradient with respect to admittance can not be obtained:

```Python
import equinox as eqx
import jax
import jax.numpy as jnp

from time_domain_differentiable.components import load_component
from time_domain_differentiable.network import Network

component = load_component(
    V_ref=2.0,
    theta_ref=0.0,
    alpha=1.0,
)

network = Network(
    G=jnp.array([[2.0]], dtype=jnp.float64),
    B=jnp.array([[3.0]], dtype=jnp.float64),
    components=((component,),),
)

voltages = jnp.array(
    [[1.2, 0.3]],
    dtype=jnp.float64,
)

states = jnp.array(
    [[[0.8, 0.4, 1.2, 0.3]]],
    dtype=jnp.float64,
)

def network_loss(module):
    i_real, i_imag = module.calculate_currents_network(
        voltages,
        use_fused=True,
    )
    return jnp.sum(i_real + 0.5 * i_imag)

def component_loss(module):
    i_real, i_imag = module.calculate_currents_components(states)
    return jnp.sum(i_real + 0.5 * i_imag)

network_grads = eqx.filter_grad(network_loss)(network)
component_grads = eqx.filter_grad(component_loss)(network)

print("grad G:", network_grads.G)
print("grad B:", network_grads.B)
print("grad Y cache:", network_grads._Y_complex_dense)

print(
    "grad component.param_vec:",
    component_grads.components[0][0].param_vec,
)
print(
    "grad packed parameters:",
    component_grads._component_param_vecs,
)
```

When using cpu, I got:
grad G: [[0.]]
grad B: [[0.]]
grad Y cache: [[1.32371591-0.21857765j]]
grad component.param_vec: [0. 0. 0.]
grad packed parameters: [[[0.20240451 0.         0.20678682]]]

When using gpu, I got:
grad G: [[1.32371591]]
grad B: [[0.21857765]]
grad Y cache: [[0.+0.j]]
grad component.param_vec: [0. 0. 0.]
grad packed parameters: [[[0.20240451 0.         0.20678682]]]
